### DISCUSSION TITLE
.NET 8 - Blazor in Module - dotnet.js file not served.

### DISCUSSION THREAD

**Discussion by websitewill** [OrchardCMS/OrchardCore#15653] [Q&A]
I am still trying to get everything to work with Blazor.
I now have a blazor component being hosted from within an existing OC module.
The blazor component is in a separate project. It all works with RenderMode Server (and ServerPrerendered).
What doesn't work is the WebAssembly mode(s).
I can see that the blazor.web.js file is being downloaded from
/_framework/blazor.web.js 

However that tries to pull this dotnet.js file and my Orchard setup is not serving that.
This file exists in the Blazor project itself - I see it in the /bin/Debug/net.8.0 folder.

What I suspect is that the way I have my OC setup, that file is not visible. Any ideas on what to do to make that also be available from /_framework/dotnet.js where it is expected?

***EDIT***

After more research, it looks as though a "normal" razor app would have routing inside if itself that is capable of serving all of the files from the _framework folder. I notice in the build output of my Orchard Module that none of the output from the Blazor web assembly project is included except for the dll and pdbs (none of the wasm, gz or js files). These would all be needed for this content to get executed in the browser. Not sure what I need to do to make this stuff available in OrchardCore. 

As a test, I added controller+index view directly into my Orchard Web project. The page loads, the blazor component renders (initially) but is still not interactive there. Seems like OC is doing something with the routing to I need to address to make that dotnet.js (and everything else that is supposed to be in _framework) available in OC.

As an additional test, I used the same Blazor project in a vanilla MVC site - it works just fine there.

TIA!

**Comment by Skrypt** [comment]
Maybe follow the example here from @ns8482e 
https://github.com/OrchardCMS/OrchardCore/pull/15647

**Reply by websitewill** [reply]
Agreed - and I have that specific branch loaded. Something not-so-obvious is different. My server-side component works but the WASM one doesn't. Still looking at how Niraj set his up to see if I can spot any obvious difference.

It's troublesome that a lot of the config stuff seems to be order-dependent and it is really hard to find good documents on what comes before what. 

If I had to guess this is something to do with routing, or allowing those static assets up from the blazor project.

@ns8482e would you at all be interested in providing some dedicated help for how you figured out what changes were needed to make this work in OC?

**Reply by Skrypt** [reply]
Ok, well it is expected that there are things to figure out still as it is a draft PR. That's the price to pay for living on the edge :wink:

**Reply by websitewill** [reply]
Agreed - and happy to contribute back any findings along the way.

**Comment by ns8482e** [comment]
Keep Blazor component in razorclasslib and add it as reference to you module

**Reply by websitewill** [reply]
It is. 

Worth noting that when I pulled your branch and ran it, I also see the errors in downloading that same dotnet.js file.
Wondering if you hit the same snag as me where things seem to work (but only server side and not via WASM)?

I have a MyModule and MyModule.Blazor project (different names, of course).
.Blazor is referenced by MyModule. MyModule is referenced in the main .Web project.

The module works (other parts of it). In the main module I have a view that declares the same Blazor component twice, once ServerPrerendered and once WebAssemblyPrerendered. Both show up in the UI, but only the first one is interactive.
![image](https://github.com/OrchardCMS/OrchardCore/assets/19608705/f5a7baeb-20a6-4e52-b11c-113095ff206a)
![image](https://github.com/OrchardCMS/OrchardCore/assets/19608705/43a2898b-1f05-4205-a608-71a78f1cee1c)

Sorry for all the redacting, but I can't quite divulge the reason for the development. :/
The first Client.OrchardModule is a standard OC module (it has existed for years).
The second redacted project is the .Blazor
The third is a .Shared - similar to your Common I think. Just some common object definitions used by both of the other projects. This is not even used yet since the PoC is only concerning the module and the .Blazor project.

**Reply by websitewill** [reply]
@ns8482e 
I came across this. Before I go too far down a rabbit hole, maybe it is the key?

https://github.com/OrchardCMS/OrchardCore/issues/11407

**Reply by websitewill** [reply]
This has to be part of the key:
https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/static-files?view=aspnetcore-8.0

But I am not quite sure (yet) what needs to be done to make this work to serve the dotnet.js + whatever files are needed for wasm from inside of the OC module (which is an MVC area, right)?

**Reply by ns8482e** [reply]
That's due to web.js- It's only download WebAssembly (dotnet.js)if it detects WebAssembly client project. 
Solution is to have an WebAssembly client project directly added to your web app and have that WebAssembly component render on your Layout.

**Reply by websitewill** [reply]
But does that mean the other stuff in Orchard (like rendering a section only if a module is enabled) is not possible.
The goal was to have this be a part of an existing module (just something that renders in an existing tag with a specific ID).

I think I am getting closer and pretty confident it is just a routing issue.
I've set the base href to "/" (also tried "~/").

Right now I am looking to see if there is something special I need to do when I build the solution.
If the Blazor project is referenced from my module project blazor end up looking for the main files in the _content folder (which doesn't seem to exist).

I've tried navigating to various files directly using the module namespace (area name) but not finding them. Currently looking into whether these blazor projects need to build as content, enbedded resoruces, copy to output, or whatever settings are required to make those files end up in the correct location of the build output for the module that references it.

**Reply by websitewill** [reply]
@ns8482e and chance you would be available/willing to take a look at what I am doing in hopes of pointing out my error?
Alternately, can you provide guidance on how to use that TextFieldPredefinedListEditor you worked with in your branch? Maybe I can debug through it and get some hint as to where my code is different.

Thanks!

**Reply by websitewill** [reply]
More follow-up. Seems like the files aren't available through my module at all.
As a test, I got rid of the BlazorWebAssembly project and included the razor/blazor components in a Components folder of the OC module itself.
Before, no matter that URL I tried, I was not able to find the dotnet.js. I was trying with the module namespace as the area (the way normal js/css files would be referenced within the module).
What I found is that the wasm stuff gets built out to the bin folder of the module. If I copy/paste those into the module's wwwroot, they still aren't available. BUT, If I set them all to Embedded resource, then I am able to load the dotnet.js (along with the .gz and .wasm files).

So, what in the world needs to happen for a component reference on a cshtml file to be able to properly reference all of those wasm files (whether from an external WebAssembly project or directly in the OC module project)?

Something funky I need to setup in the module-specific startup.cs file?
What makes OC serve files found in the bin/Debug/net8.0/wwwroot folder (and not just from embedded resources)?
Or do I need to do some build/packaging magic in order to make this all work?
